src/Sarif.Viewer.VisualStudio/Tags/TextViewCaretListener.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            SnapshotPoint caretSnapshotPoint = caretPosition.BufferPosition;
+
+            NormalizedSnapshotSpanCollection normalizedSnapshotSpanCollection = new NormalizedSnapshotSpanCollection(new SnapshotSpan(caretSnapshotPoint, 1));
+
+            List<ISarifLocationTag> tagsCaretIsCurrentlyIn = this.tagger.GetTags(normalizedSnapshotSpanCollection).
+                Where(tag => tag.Tag is ISarifLocationTag).
+                Select(tag => tag.Tag as ISarifLocationTag).
+                ToList();
+
+            if (this.previousTagsCaretWasIn != null && tagsCaretIsCurrentlyIn.SequenceEqual(this.previousTagsCaretWasIn, this.textMarkerTagCompaerer))
+            {
+                return;
+            }
+
+            IEnumerable<ISarifLocationTag> tagsToNotifyOfCaretEnter = tagsCaretIsCurrentlyIn.Where(tagCaretIsCurrentlyIn => this.previousTagsCaretWasIn == null || !this.previousTagsCaretWasIn.Contains(tagCaretIsCurrentlyIn, this.textMarkerTagCompaerer));
+            IEnumerable<ISarifLocationTag> tagsToNotifyOfCaretLeave = this.previousTagsCaretWasIn?.Except(tagsToNotifyOfCaretEnter, this.textMarkerTagCompaerer);
+
+            // Start an update batch in case the notifications cause a series of changes to tags. (Such as highlight colors).
+            foreach (ISarifLocationTag tagToNotify in tagsToNotifyOfCaretEnter)
+            {
+                tagToNotify.NotifyCaretEntered();
+            }
+
+            if (tagsToNotifyOfCaretLeave != null)
+            {
+                foreach (ISarifLocationTag tagToNotify in tagsToNotifyOfCaretLeave)
+                {
+                    tagToNotify.NotifyCaretLeft();
+                }
+            }
+
+            this.previousTagsCaretWasIn = tagsCaretIsCurrentlyIn;
+        }
+
+        private void TextView_Closed(object sender, EventArgs e)
+        {
+            this.UnsubscribeFromEvents();
+        }
+
+        private void UnsubscribeFromEvents()
+        {
+            this.textView.Closed -= this.TextView_Closed;
+            this.textView.LayoutChanged -= this.TextView_LayoutChanged;
+            this.textView.Caret.PositionChanged -= this.Caret_PositionChanged;
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (this.isDisposed)
+            {
+                return;
+            }
+
+            this.isDisposed = true;
+            if (disposing)
+            {
+                this.UnsubscribeFromEvents();
+
+                using (ExistingListenersLock.EnterWriteLock())
+                {
+                    ExistingListeners.Remove(this.textView);
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        private class TextMarkerTagCompaerer : IEqualityComparer<ISarifLocationTag>
```

Comment:
>TextMarkerTagCompaerer [](start = 22, length = 22)

Comment why this is needed and what it's purpose is.

(This is used to make sure we don't send enter\leave notifications to the same tags above - the previous code was doing this, and slowing down the editor).